### PR TITLE
Added basic lockfile generation, fixes #4

### LINKED ISSUE #4: Generate the package list file after installing a module

### PR DESCRIPTION
**Pull request by MrNaif2018**
Fixes #4
This PR adds basic lockfile generation.
It is more of a draft as I have some questions (and found bugs) about the implementation:

- How should deps be listed? In tarballs downloaded, I don't see any file specifying dependencies (:
- Are there any formatting guidelines set up for this repo? If not, I can help with setting up black+isort+CI (:
- When using `--login`, it adds to `.gitignore`, which probably isn't needed anymore, should I remove that?
- Github access token was not used before, so I added loading from `.env` file
- Currently lockfile will only have one dependency in it as nuclear allows installing only one package
- Should lockfile be local like I did, or global? Should it append or rewrite?
- When installing same package twice it failed because of `makedirs` `exist_ok=False` and `xb` open mode. Is it expected? Why disallowing re-installing same package? I have removed it in the PR
- Should `get_latest_commit` default to `master` or fetch data from repository metdata? (one additional API call)
- When installing from github with no releases, I used latest commit as version

Sample `nuclear.lock` contents:
```
### DO NOT EDIT: THIS IS AN AUTOGENERATED FILE ###
### nuclear.lock version 1.0.0 ###

ArkScript-lang/std v0.0.1 https://api.github.com/repos/ArkScript-lang/std/tarball/v0.0.1 35ef4104539a5765395f0403f5e690573b5fe95a64253bdf6566d5462c11762a
```

### CODE REVIEW

**Comment by SuperFola** [comment]
Hi there, thanks for your PR

Let me address your questions:

> How should deps be listed? In tarballs downloaded, I don't see any file specifying dependencies (:

We don't have any yet, that would be the job of the lock file. Some people would think that there should be a packages.json and a packages.lock, as npm is doing ; I'm coming from Python where those don't exist and we work only with a requirements.txt and I find that easier to use. Maybe I'm wrong and there are advantages to have both, I'm always open to discussion!

> Are there any formatting guidelines set up for this repo? If not, I can help with setting up black+isort+CI (:

Not yet, I started programming while keeping the PEP8 in my head, and using 4 spaces instead of tabs ^^'. When talking about CI you think of think Actions of something external?

> When using --login, it adds to .gitignore, which probably isn't needed anymore, should I remove that?

I see that as a way to prevent token leaking since this file can be auto generated, the user can forget to put it in the .gitignore themself (I agree that people won't forget to put their .env file in the .gitignore if they write it themselves).

> Github access token was not used before, so I added loading from .env file

By default the requests were made in anonymous mode, I forgot that use case where the user was previously logged in, thanks.

> Currently lockfile will only have one dependency in it as nuclear allows installing only one package

We can install multiple packages as long as they are all different (source, author, name, version).

> Should lockfile be local like I did, or global? Should it append or rewrite?

It should be local to a project and append when a new dependency is added, while the packages should be extracted to a global place.

> When installing same package twice it failed because of makedirs exist_ok=False and xb open mode. Is it expected? Why disallowing re-installing same package? I have removed it in the PR

That wasn't expected, I must have missed it when I tested the previous PR 😅 Indeed a package could be reinstalled, either the same version, or a newer/older one!

> Should get_latest_commit default to master or fetch data from repository metdata? (one additional API call)

Since GitHub is pushing people to use other branch names (like "main" or "dev"), we could fetch data from the repo metadata, I think there is a way to kill two birds in one stone, ie retrieve versions information AND repository metadata.

> When installing from github with no releases, I used latest commit as version

That was also intended: if the user provides no version number, and there isn't any release, it's fetching the latest commit

File: nuclear/utils/lockfile.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+LOCKFILE_VERSION = "1.0.0"
+
+def save(user, repo, version, url, sha256):
```
Comment (SuperFola) [inline nuclear/utils/lockfile.py]: Could you add type information here as well to be consistent with the rest of the code? :D

**Comment by MrNaif2018** [comment]
> We don't have any yet, that would be the job of the lock file. Some people would think that there should be a packages.json and a packages.lock, as npm is doing ; I'm coming from Python where those don't exist and we work only with a requirements.txt and I find that easier to use. Maybe I'm wrong and there are advantages to have both, I'm always open to discussion!

I think we should have best of both worlds-a requirements.txt file or similar, plus lockfile. It might be similar to what pipenv does, but here it will be better (:

> Not yet, I started programming while keeping the PEP8 in my head, and using 4 spaces instead of tabs ^^'. When talking about CI you think of think Actions of something external?

I suggest you to try black+isort formatting set up for consistency, ease of use and automatic code formatting checking on CI. By CI (Continuous Integration) I mean either CircleCI or github actions.



> I see that as a way to prevent token leaking since this file can be auto generated, the user can forget to put it in the .gitignore themself (I agree that people won't forget to put their .env file in the .gitignore if they write it themselves).

I know, but `.env` file is already in .gitignore, there is no need to re-add it.



> It should be local to a project and append when a new dependency is added, while the packages should be extracted to a global place.

And what if user decides to install same package again? It should find the entry and re-write, or append?



> Since GitHub is pushing people to use other branch names (like "main" or "dev"), we could fetch data from the repo metadata, I think there is a way to kill two birds in one stone, ie retrieve versions information AND repository metadata.

Not sure if it's possible, but I will try to find a way

**Comment by SuperFola** [comment]
> I think we should have best of both worlds-a requirements.txt file or similar, plus lockfile. It might be similar to what pipenv does, but here it will be better (:

After looking up how this would come in handy, I agree!

> I suggest you to try black+isort formatting set up for consistency, ease of use and automatic code formatting checking on CI. By CI (Continuous Integration) I mean either CircleCI or github actions.

Thanks, I'll look that up

> I know, but .env file is already in .gitignore, there is no need to re-add it.

Correction, `.env` is in the .gitignore of this project, but nuclear can be called from everywhere (in the end it should be installed globally for the end user, for example in /usr/bin/nuclear) thus we will have to deal with the .env file of the (possible) git repository nuclear will be ran from.

> And what if user decides to install same package again? It should find the entry and re-write, or append?

If it's the exact same package (source, user, name, version), then just try to fetch it again from the URL in the lock file (we can assume that the download failed somehow) ; if it's a downgrade/upgrade (same source, user, name), then re-write the entry in the lockfile.

**Comment by MrNaif2018** [comment]
> Correction, .env is in the .gitignore of this project, but nuclear can be called from everywhere (in the end it should be installed globally for the end user, for example in /usr/bin/nuclear) thus we will have to deal with the .env file of the (possible) git repository nuclear will be ran from.

Ah ok. Then it should add to gitignore by checking if it's there first already. As when running it multiple times I got
`.env.env.env.env` in my .gitignore

**Comment by SuperFola** [comment]
> Ah ok. Then it should add to gitignore by checking if it's there first already. As when running it multiple times I got
.env.env.env.env in my .gitignore

My bad, I forgot about that being possible

**Comment by MrNaif2018** [comment]
Sorry for not finishing this before!
I have made the following changes:
- Adding .env to .gitignore should be smarter now, not adding it twice
- Getting latest commit is done in one request when a token is available - using github graphql api - otherwise a warning is displayed and two requests are used
- Improved lockfile generation:
How it works:
First, it collects all metadata about packages it is going to install
Afterwards, it tries reading from a lockfile if possible, and parses all already installed packages.
Then it adds/overrides newly installed packages
Then lockfile is overwritten with new package data.
Why not just appending? Because with this approach of re-writing full file it would be easier to perform lockfile migration to a newer version.

My `nuclear.lock`:
```
### DO NOT EDIT: THIS IS AN AUTOGENERATED FILE ###
### nuclear.lock version 1.0.0 ###

ArkScript-lang/std v0.0.1 https://api.github.com/repos/ArkScript-lang/std/tarball/v0.0.1 35ef4104539a5765395f0403f5e690573b5fe95a64253bdf6566d5462c11762a
ArkScript-lang/rfc d7414affb6f03c446518fc4550c735a15e8d9405 https://api.github.com/repos/ArkScript-lang/rfc/tarball 0a5266bde783592f437900c5b2d9a57d430be2335d8150c26c169e89fc817bc2
```

**Review by SuperFola — APPROVED** [review]

**Comment by SuperFola** [comment]
Looks good to me, tested and working, that's awesome!

Thanks a lot for your contribution

**Pull request merged**